server/src/main/java/org/elasticsearch/health/HealthService.java

```diff
@@ -80,30 +98,96 @@ public List<HealthIndicatorResult> getHealth(@Nullable String indicatorName, boo
         Stream<HealthIndicatorService> filteredIndicators = healthIndicatorServices.stream()
             .filter(service -> indicatorName == null || service.name().equals(indicatorName));
 
-        Stream<HealthIndicatorResult> filteredIndicatorResults;
         if (clusterHealthIsObtainable) {
-            // Calculate remaining indicators
-            filteredIndicatorResults = filteredIndicators.map(service -> service.calculate(explain));
+            if (HealthNode.isEnabled()) {
+                try {
+                    client.execute(FetchHealthInfoCacheAction.INSTANCE, new FetchHealthInfoCacheAction.Request(), new ActionListener<>() {
+                        @Override
+                        public void onResponse(FetchHealthInfoCacheAction.Response response) {
+                            HealthInfo healthInfo = response.getHealthInfo();
+                            combineResultsAndNotifyListener(
```

Comment:
See what you think of what I've done. It's hard to pull it all apart because the indicatorName is used for validation, and we need the listener when validation fails. I've separated out the filtering of preflight results by indicator name and the combining of results though.